### PR TITLE
Add ConsensusV2 object support to `sui-indexer{,alt}`

### PR DESCRIPTION
**Pull request by aschran**
Updates `sum_coin_balances` and `wal_coin_balances` tables to add a column for `coin_owner_kind`, so that balances can be tracked separately for fastpath vs. consensus coins.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [x] Indexer: Adds `coin_owner_kind` column to coin balances tables. 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-indexer-alt/migrations/2024-10-28-144002_sum_coin_balances/up.sql
Comment (amnn) [inline crates/sui-indexer-alt/migrations/2024-10-28-144002_sum_coin_balances/up.sql:10]: Yep, feel free to change it here.

File: crates/sui-indexer-alt/src/handlers/sum_obj_types.rs
Comment (amnn) [inline crates/sui-indexer-alt/src/handlers/sum_obj_types.rs:102]: I thought that for objects we wanted to treat consensus-owned objects as address-owned if they the authenticator maps to a single address, which means no new kind here. (Then there was a future question of how we handle party objects which we kind of punted on, but the options varied from treating them like shared objects, to having some separate category for jointly-owned objects).

File: crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql
Comment (amnn) [reply crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql:2]: Also, we are going to need to update all the indices on all these tables -- wherever they mentioned `owner_id` we now need to prefix it with an `owner_kind`.

File: crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs
Comment (amnn) [inline crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs:108]: We should use a different enum here to the one we use for objects in general, because while we don't want to distinguish consensus-managed single-owned objects from address-owned objects, we do want to do that for coins.

File: crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs
Comment (amnn) [inline crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs:109]: Just noting my momentary confusion that this doesn't return an `Option<...>`!

**Comment by lxfind** [comment]
There is now also a obj_info pipeline

File: crates/sui-indexer-alt/src/handlers/sum_obj_types.rs
Comment (aschran) [reply crates/sui-indexer-alt/src/handlers/sum_obj_types.rs]: updated as per huddle

File: crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql
Comment (aschran) [reply crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql]: added index updates

File: crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs
Comment (aschran) [reply crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs]: created a new enum

File: crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs
Comment (aschran) [reply crates/sui-indexer-alt/src/handlers/sum_coin_balances.rs:109]: yes, i put a comment on the function to clarify in another PR

File: crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql
Comment (amnn) [inline crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql:5]: I'm thinking of some ways to make this easier to land without causing the indexer to hang when it starts, and I think I have an idea:

- We can get postgres to drop and create these indices concurrently, which will mean they won't hang other things that are going on at the same time (this will allow us to run the migrations proactively while the indexer is running), 
- and then turn them into `IF EXISTS`/`IF NOT EXISTS` style statements so that they become idempotent, (which means after we've run it proactively, when the migration runs, it knows not to drop and regenerate the indices),
- and finally, we need to give the new index a new name so that `IF EXISTS`/`IF NOT EXISTS` would recognise it as different from the old index.

The statements for creating and dropping the indices would look like this then:

```suggestion
DROP INDEX CONCURRENTLY IF EXISTS sum_coin_balances_owner_type;
CREATE INDEX CONCURRENTLY IF NOT EXISTS sum_coin_balances_owner_kind_and_id;
```

But you will also need to split it up over separate migration scripts (because each script is its own transaction and postgres won't like it if you combine non-blocking concurrent operations with other work in a transaction). I've had to do this once before in #19543 so that can serve as a template (there's an extra TOML file that needs to go in each migration folder to tell diesel to not batch up migrations into a transaction as well).

(ditto for the wal tables below).

**Review by amnn — COMMENTED** [review]
This change now looks good to me! cc @lxfind though as he is currently working on the `sum_coin_balances` and `sum_obj_types` pipelines (or their equivalents), so we may need to incorporate changes there.

I've also left a comment on how to make the index changes go through smoothly. Once those changes are in, we can coordinate the land and proactively running the migration so everything keeps working smoothly.

File: crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql
Comment (aschran) [reply crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql]: I did this, let me know if it looks correct. when I run `generate_schema.sh` locally it fails with an error 
```
2024-12-05 10:37:44.702 EST [78369] ERROR:  DROP INDEX CONCURRENTLY cannot run inside a transaction block
```

But I have the `metadata.toml` files there as you suggested - is anything else missing?

Comment (amnn) [reply crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql]: The issue is that the migrations need to be split up even further -- each `DROP INDEX` and `CREATE INDEX` needs to be in their own script (because they can't mingle with each other either).

Comment (aschran) [reply crates/sui-indexer-alt/migrations/2024-12-02-185822_add_coin_balance_owner_kind/up.sql]: Thanks, done

**Review by amnn — APPROVED** [review]
LGTM! @aschran, once you have landed this, please ping the Data Platform slack channel, so someone (probably myself or @lxfind) can run the migrations proactively, so we can avoid the next deploy hanging indefinitely.

**Pull request merged**
